dsa.go

```diff
@@ -93,55 +93,47 @@ func (c *Context) GenerateDSAKeyPairWithLabel(id, label []byte, params *dsa.Para
 		return nil, errClosed
 	}
 
-	if err := notNilBytes(id, "id"); err != nil {
-		return nil, err
-	}
-	if err := notNilBytes(label, "label"); err != nil {
+	template, err := NewAttributewithLabel(id, label)
+	if err != nil {
 		return nil, err
 	}
-
-	return c.generateDSAKeyPair(id, label, params)
+	return c.GenerateDSAKeyPairWithAttributes(template, template.Copy(), params)
 }
 
-// generateDSAKeyPair creates a DSA private key on the token.
-func (c *Context) generateDSAKeyPair(id, label []byte, params *dsa.Parameters) (k *pkcs11PrivateKeyDSA, err error) {
+// GenerateDSAKeyPairWithAttributes creates a DSA key pair on the token. Additional attributes from public and
+// private will be merged into the default templates used when generating keys. Where conflicts occur, the user
+// supplied attributes will win.
+func (c *Context) GenerateDSAKeyPairWithAttributes(public, private *AttributeSet, params *dsa.Parameters) (k *pkcs11PrivateKeyDSA, err error) {
```

Comment:
The return values should be _unnamed_ and using the public interface type, e.g. `(Signer, error)`. This same comment applies to the other `...WithAttributes` functions.

Reply:
I just noticed this same issue when I updated my codebase to use a version of this PR. Yes, I will change these to return the `Signer` interface instead.

Reply:
I switched the return type to `Signer` but left it named as the bulk of the logic is within the anonymous function passed to `withSession()` and the named argument allows the value to be easily returned from the outer function.